driver-scala/src/main/scala/org/mongodb/scala/model/search/SearchFacet.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.mongodb.scala.model.search
+
+import com.mongodb.annotations.Beta
+import com.mongodb.client.model.search.{ SearchFacet => JSearchFacet }
+import org.mongodb.scala.bson.BsonDocument
+import org.mongodb.scala.bson.conversions.Bson
+
+import java.time.Instant
+import collection.JavaConverters._
+
+/**
+ * A facet definition for [[FacetSearchCollector]].
+ *
+ * @see [[https://www.mongodb.com/docs/atlas/atlas-search/facet/#facet-definition Facet definition]]
+ * @since 4.6
+ */
+@Beta
+object SearchFacet {
+
+  /**
+   * Returns a `SearchFacet` that allows narrowing down search results based on the most frequent
+   * values in the BSON `String` field specified by the `path`.
+   *
+   * @param name The facet name.
+   * @param path The path to facet on.
+   * @return The requested `SearchFacet`.
+   * @see [[https://www.mongodb.com/docs/atlas/atlas-search/facet/#string-facets String facet definition]]
+   */
+  def stringFacet(name: String, path: FieldSearchPath): StringSearchFacet = JSearchFacet.stringFacet(name, path)
+
+  /**
+   * Returns a `SearchFacet` that allows determining the frequency of
+   * BSON `32-bit integer` / `64-bit integer` / `Double` values
+   * in the search results by breaking the results into separate ranges.
+   *
+   * @param name The facet name.
+   * @param path The path to facet on.
+   * @param boundaries Bucket boundaries in ascending order. Must contain at least two boundaries.
+   * The supported `Number` types are the same as those supported by org.bson.BasicBSONEncoder.putNumber`.
+   * @return The requested `SearchFacet`.
+   * @see [[https://www.mongodb.com/docs/atlas/atlas-search/facet/#numeric-facets Numeric facet definition]]
+   */
+  def numberFacet(name: String, path: FieldSearchPath, boundaries: Iterable[_ <: Number]): NumericSearchFacet =
```

Comment:
Using `Iterable[_ <: Number]` here is a problem, because `Number` is not a Scala type. As a result, users have to write `List[Number](10, 20, 30)` instead of writing `List(10, 20, 30)`, for example.

A similar problem is with `Iterable[Instant]` (`Instant` is not specified as an upper boundary because this class is `final`).

@rozza Could you please help with figuring out what are the right types to use in Scala, and how to convert from them to Java's `Iterable<? extends Number>` and `Iterable<Instant>`?

Reply:
Just use: `Iterable[Number]` and then the converter will handle the types correctly eg:

```
println(SearchFacet.numberFacet("a", fieldPath("numberFieldName"), Seq(10, 20.0, 30L)))
```

Also: Instant seems the appropriate type in scala :+1:

Reply:
> Instant seems the appropriate type in scala

Thanks!, at least that part is done :)

> Just use: Iterable[Number]

In Java `Iterable<Double>` is not of type `Iterable<Number>`, but I would like to allow users to pass `Iterable<Double>` as well as any other `Iterable<N>` where `N extends Number`. This can be done in one of the following ways:

1. `<N extends Number> NumericSearchFacet numberFacet(..., Iterable<N> boundaries)`
2. `NumericSearchFacet numberFacet(..., Iterable<? extends Number> boundaries)`, which is almost exactly the asme thing as the first option, only instead of having of the type parameter `N`, we don't have any name and cannot refer to this type parameter in the body of the method.

Based on https://docs.scala-lang.org/tour/upper-type-bounds.html, my understanding is that in Scala 2 `Iterable[_ <: Number]` does exactly the same thing as #2 in Java. If that is correct, then using `Iterable[Number]` is not equivalent to `Iterable[_ <: Number]`, and is unjustifiably restrictive.

If I use `Iterable[Number]`, then `SearchFacet.numberFacet("a", SearchPath.fieldPath("numberFieldName"), Seq(10, 20.0, 30L))` indeed compiles. However, the following code does not compile:
```scala
val boundaries: Seq[scala.Double] = Seq(1.0, 2.0)
// `var boundaries = Seq(1.0, 2.0)` also fails with the same error
val facet: SearchFacet = SearchFacet.numberFacet("facetName", SearchPath.fieldPath("fieldName"), boundaries)
```
```
type mismatch;
 found   : Seq[Double]
 required: Iterable[Number]
```

However, the above code does not compile even if I use `Iterable[_ <: Number]`, which implies that "transparent" convertors from `collection.JavaConverters._` are not exactly "transparent" and break type expectations. Because of this, I will use `Iterable[Number]`, but the fact that users are not allowed to pass `Seq[scala.Double]`/`Seq[scala.Int]`/etc. is disheartening.

Reply:
There are quirks to scala and its compiler :)

Why is removing bounds helpful? Normally it wouldn't be if you were using Scala types - it would be the correct thing to do. However, `Number` is not a Scala type and it has its own Numeric types.

So usage becomes the key issue here and making it as simple as possible for the user.

```
val boundaries: Seq[Double] = Seq(10, 20.0, 30L)
```

Will never work as `scala.Double` is nothing to do with `Number`. So you force the user to cast to a `Number` type. Where as not setting the type inline will work as it allows the compiler to inspect the type and automatically cast to Java `Number`.

By restricting the types, you allow users boundaries inline eg:

```
SearchFacet.numberFacet("a", fieldPath("numberFieldName"), Seq(10, 20.0, 30L))
```

You are getting out of the users way. Other wise by setting the bounds you are forcing the user to declare their types, either inline or by values eg:

```
val boundaries: Seq[Number] = Seq(10, 20.0, 30L)
SearchFacet.numberFacet("a", fieldPath("numberFieldName"), boundaries)

val boundaries: Seq[java.lang.Double] = Seq(10, 20.0, 30L)
SearchFacet.numberFacet("a", fieldPath("numberFieldName"), boundaries)
```

The downside is you also force inline usage to declare the type:

```
SearchFacet.numberFacet("a", fieldPath("numberFieldName"), Seq[NUMBER TYPE NEEDED](10, 20.0, 30L))

```

So bounds are technically correct, as Scala has its own number types, it actually gets in the way
for users in all scenarios. So thats why I recommend not setting them.